### PR TITLE
Removed avro dependency for StatsManager

### PR DESCRIPTION
**Pull request by xunyin8**
Removed the dependency on Avro for StatsManager. Avro 1.4.1 is required for another component to function and downgrading to 1.4.1 is difficult since there is no convenient way of performing Avro code generation with 1.4.1. The data/model objects (StatsSnapshot, StatsHeader, and StatsWrapper) will become generic java classes.
For convenience, StatsManager will now serialize and publish the stats to JSON via Jackson's ObjectMapper (existing dependency). If needed in the future, we can also introduce our own serialization/deserialization.

`./gradlew build && ./gradlew test` success
coding style applied

### CODE REVIEW

File: ambry-api/src/main/java/com.github.ambry/clustermap/ResourceStatePolicy.java [hunk 1]
```diff
@@ -50,6 +50,5 @@
    * Should be called if the caller knows outside of the policy that the resource is up.
    */
   void onHardUp();
-
```
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/clustermap/ResourceStatePolicy.java]: intended change?

File: ambry-api/src/main/java/com.github.ambry/server/StatsHeader.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * A model object that contains metadata information about some reported stats. For example, the kind of stats that is
+ * being reported, timestamp and etc.
+ */
+public class StatsHeader {
+  public enum StatsDescription {
+    QUOTA
+  }
+
+  private StatsDescription description;
```
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/server/StatsHeader.java]: all `final`

File: ambry-api/src/main/java/com.github.ambry/server/StatsHeader.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+package com.github.ambry.server;
+
+import java.util.List;
+
+
+/**
+ * A model object that contains metadata information about some reported stats. For example, the kind of stats that is
+ * being reported, timestamp and etc.
+ */
+public class StatsHeader {
```
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/server/StatsHeader.java]: does the class and `enum` need to be public (v/s package private)? (is this for the jackson object mapper?)

File: ambry-api/src/main/java/com.github.ambry/server/StatsHeader.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  StatsHeader(StatsDescription description, long timestamp, int storeContacted, int storeResponded,
+      List<String> unreachableStores) {
+    this.description = description;
+    this.timestamp = timestamp;
+    this.storeContacted = storeContacted;
+    this.storeResponded = storeResponded;
+    this.unreachableStores = unreachableStores;
+  }
+
+  public StatsDescription getDescription() {
```
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/server/StatsHeader.java]: functions need to be public?  (is this for the jackson object mapper?)

File: ambry-api/src/main/java/com.github.ambry/server/StatsSnapshot.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  private Map<String, StatsSnapshot> subtree;
+
+  /**
+   * Walk through two {@link StatsSnapshot} and check if they are equal. Two {@link StatsSnapshot} are considered equal
+   * if they have the same value and subtree mapping at each level.
+   * @param a the {@link StatsSnapshot} to be compared with b
+   * @param b the {@link StatsSnapshot} to be compared with a
+   * @return
+   */
+  public static boolean isEqual(StatsSnapshot a, StatsSnapshot b) {
```
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/server/StatsSnapshot.java]: why not override the `equals()` function instead? If that is applicable, you can even autogenerate the code inside intellij.

File: ambry-api/src/main/java/com.github.ambry/server/StatsSnapshot.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+  public long getValue() {
+    return value;
+  }
+
+  public Map<String, StatsSnapshot> getSubtree() {
+    return subtree;
+  }
+
+  public void setValue(long value) {
```
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/server/StatsSnapshot.java]: are these setters used anywhere?

File: ambry-api/src/main/java/com.github.ambry/clustermap/ResourceStatePolicy.java (hunk 1 above)
Comment (xunyin8) [inline ambry-api/src/main/java/com.github.ambry/clustermap/ResourceStatePolicy.java]: not intended but it's from applying coding style format.

File: ambry-api/src/main/java/com.github.ambry/server/StatsSnapshot.java (hunk 6 above)
Comment (xunyin8) [inline ambry-api/src/main/java/com.github.ambry/server/StatsSnapshot.java]: Yes, used during aggregation in StatsManager

File: ambry-api/src/main/java/com.github.ambry/server/StatsHeader.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    QUOTA
+  }
+
+  private final StatsDescription description;
+  private final long timestamp;
+  private final int storeContacted;
+  private final int storeResponded;
+  private final List<String> unreachableStores;
+
+  StatsHeader(StatsDescription description, long timestamp, int storeContacted, int storeResponded,
```
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/server/StatsHeader.java]: minor: `storesResponded` or `storesRespondedCount`? (same with contacted)

File: ambry-api/src/main/java/com.github.ambry/server/StatsSnapshot.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+ * the subtree field will contain a mapping of accountIds to {@link StatsSnapshot}. Each mapped {@link StatsSnapshot}'s
+ * value will be the total valid data size for all containers that belong to the account that is mapped with. The mapped
+ * {@link StatsSnapshot}'s subtree will contain all the containers in the account that is mapped with. At the leaf level
+ * {@link StatsSnapshot}'s subtree will be null.
+ */
+public class StatsSnapshot {
+  private long value;
+  private Map<String, StatsSnapshot> subtree;
+
+  public StatsSnapshot(Long value, Map<String, StatsSnapshot> subtree) {
```
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/server/StatsSnapshot.java]: minor: rename to `subMaps`?

**Comment by xunyin8** [comment]
Updated, `./gadlew build && ./gradlew test` success on both Mac and Linux.

**Pull request merged**